synctest/run_app.go

```diff
@@ -0,0 +1,244 @@
+			return err
+		}
+
+		for _, v := range accounts {
+			_, err := cmd.AddGenAccount(v, "1000000000000000utia", home)
+			if err != nil {
+				return err
+			}
+		}
+
+		gen, err := os.Open(fmt.Sprintf("%s/config/genesis.json", home))
+		if err != nil {
+			return err
+		}
+
+		bt, err := ioutil.ReadAll(gen)
+		if err != nil {
+			return err
+		}
+
+		_, err = client.Publish(ctx, testkit.InitialGenenesisTopic, string(bt))
+		if err != nil {
+			return err
+		}
+
+		runenv.RecordMessage("Orchestrator has sent initial genesis with accounts")
+	} else {
+		initGenCh := make(chan string)
+		sub, err := client.Subscribe(ctx, testkit.InitialGenenesisTopic, initGenCh)
+		if err != nil {
+			return err
+		}
+		select {
+		case err = <-sub.Done():
+			if err != nil {
+				return err
+			}
+		case initGen := <-initGenCh:
+			err = os.WriteFile(fmt.Sprintf("%s/config/genesis.json", home), []byte(initGen), 0777)
+			if err != nil {
+				return err
+			}
+		}
+		runenv.RecordMessage("Validator has received the initial genesis")
+	}
+
+	_, err = cmd.SignGenTx(keyringName, "5000000000utia", "test", chainId, home)
+	if err != nil {
+		return err
+	}
+
+	fs, err := os.ReadDir(fmt.Sprintf("%s/config/gentx", home))
+	if err != nil {
+		return err
+	}
+	// slice is needed because of auto-gen gentx-name
+	for _, f := range fs {
+		gentx, err := os.Open(fmt.Sprintf("%s/config/gentx/%s", home, f.Name()))
+		if err != nil {
+			return err
+		}
+
+		bt, err := ioutil.ReadAll(gentx)
+		if err != nil {
+			return err
+		}
+
+		_, err = client.Publish(ctx, testkit.GenesisTxTopic, string(bt))
+		if err != nil {
+			return err
+		}
+
+	}
+
+	genTxCh := make(chan string)
+	sub, err := client.Subscribe(ctx, testkit.GenesisTxTopic, genTxCh)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < runenv.TestGroupInstanceCount; i++ {
+		select {
+		case err = <-sub.Done():
+			if err != nil {
+				return err
+			}
+		case genTx := <-genTxCh:
+			if !strings.Contains(genTx, accAddr) {
+				err := ioutil.WriteFile(fmt.Sprintf("%s/config/gentx/%d.json", home, i), []byte(genTx), 0777)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	_, err = cmd.CollectGenTxs(home)
+	if err != nil {
+		return err
+	}
+
+	configPath := filepath.Join(home, "config", "config.toml")
+	err = appkit.ChangeNodeMode(configPath, "validator")
+	if err != nil {
+		return err
+	}
+
+	err = appkit.ChangeRPCServerAddress(configPath, net.ParseIP("0.0.0.0"))
+	if err != nil {
+		return err
+	}
+
+	runenv.RecordMessage("publishing app-validator address")
+	ip, err := initCtx.NetClient.GetDataNetworkIP()
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Publish(
+		ctx,
+		testkit.AppNodeTopic,
+		&testkit.AppNodeInfo{
+			ID: int(initCtx.GroupSeq),
+			IP: ip,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	runenv.RecordMessage("starting........")
+	go cmd.StartNode(home)
+
+	// wait for a new block to be produced
+	// RPC is also being initialized...
+	time.Sleep(30 * time.Second)
```

Comment:
I believe 30 seconds is too much and potential will build up into a vacation amount of time wasted waiting for testground tests to finish, lol. Check out [the code here](https://github.com/celestiaorg/celestia-core/blob/v0.35.x-celestia/rpc/test/helpers.go#L91) for reference on how to know when RPC is started.

Reply:
The RPC is starting but still I get no proper response back and only error codes from the endpoint. So I propose we leave this until better times and capacity to deep dive into the reason on why even after the rpc is started - there is still delay to get the response messages properly